lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -3619,7 +3651,7 @@ CheckedCastKind TypeChecker::typeCheckCheckedCast(Type fromType,
   // If the destination type can be a supertype of the source type, we are
   // performing what looks like an upcast except it rebinds generic
   // parameters.
-  if (fromType->isBindableTo(toType))
+  if (toType->isBindableTo(fromType))
```

Comment:
Could you explain why this got flipped? Looking at the comment above, it seems like the original one was correct... Is the comment out of sync?

Reply:
For what I understood it was exactly the other way around
```swift 
func SR13035<SomeError: SR13035Error>(_ mockResult: Result<String, ChildError>, _: Result<String, SomeError>) {
      let _ = mockResult as? Result<String, SomeError> // Ok
}
```
So here to verify what the comment was saying, we should check if we can bind `Result<String, SomeError>` to `Result<String, ChildError>`, so that is why I inverted :)

Reply:
My reading was we need to check if we can bind `Result<String, ChildError>` (source type, i.e. `fromType`) to `Result<String, SomeError>` (destination type, i.e. `toType`) since the latter can be a supertype of the former. 🤔 So going from `Result<String, ChildError>` to `Result<String, SomeError>` looks like an upcast except it rebinds generic parameters.

Reply:
Yes, exactly. I think that is a bit confusing because of the `isBindable` function. The way it was it was failing to check that for this situation.

Reply:
Let me take a more detailed look at this and either to clarify comments or change the implementation :)

Reply:
Seems like the is bindable only trying to substitute the generic archetype from the this -> argument type. In this case toType (SomeError generic archetype) -> to the generic argument of type `ChildError`.

Reply:
What about the other way around, e.g `Result<String, SomeError>` to `Result<String, ChildError>`? In both cases it seems like we're doing a dynamic check for whether `SomeError` is `ChildError` and performing the cast, so it seems like both cases should be treated as a `ValueCast`.

Reply:
Humm ... I have to check, but I'm not sure this will be handled here, let me check this :)

Reply:
@hamishknight You were right we need both cases :)

Reply:
Already adjusted and test case added

Reply:
Thanks!